decision-records/2024-11-07--dashboard-title-collision.md

```diff
@@ -0,0 +1,85 @@
+# Dashboard title collision in Grafana
+**Date:** 2024-11-07<br/>
+**Authors:** @lucabello @sed-i
+
+
+## Context and Problem Statement
+Dashboard json includes keys such as in the following example:
+```bash
+$ cat metrics-dashboard.json | jq '{"title", "uid", "version", "tags"}'
+{
+  "title": "Loki Operator Overview",
+  "uid": "loki",
+  "version": 1,
+  "tags": [
+    "loki"
+  ]
+}
+```
+
+Grafana [documented](https://grafana.com/docs/grafana/latest/administration/provisioning/#reusable-dashboard-urls) that dashboards should not have the same *title* within a folder, or the same *uid* within
+installations, which causes some problems in different scenarios.
+
+This ADR addresses **title** collisions.
+
+Dashboards from different charms may happen to have the same title:
+
+`loki(dashboard "Overview") <--> grafana <--> tempo(dashboard "Overview")`
+
+If both charms have a dashboard simply named "Overview", we'll have problems.
+We should not put charm authors in a position where they need to be mindful of an entire hypothetical deployment and try
+to avoid potential title collisions.
+
+
+## Requirements
+1. Dashboards must not have the same *title* within a folder.
+2. Juju admins, not charm authors, should have the final say on how dashboards are organized in sub-folders.
+
+## Decision
+(2/P).
+
+## Alternatives considered
+Deduplication can be addressed on the provider (charm with dashboards) or the requrier side (grafana charm).
+In the following subsections, `P` and `R` stand for provider and requirer.
+
+### (1/R) Place dashboards in folders by charm name
+
+We can create a folder per charm (using the charm name, not the app name) and put the dashboard in there.
+This is possible because the charm name is already part of the relation data schema (Appendix 1).
+
+Under the top level folders named after the charm we'd preserve the tree structure the charm author had in place for the
+dashboards.
+
+### (2/P) Let charm authors or admins specify the folder name under which grafana should place its dashboards
+
+This alternative relies on charm authors configuring their charms accordingly. The default behavior would be either the "General" folder (as it is right now) or a folder named after the charm name.
+We could add a constructor arg for the name of the folder. Charm authors can choose whether to expose it as a config option. The constructor arg(s) could offer the following functionality:
```

Comment:
> We could add a constructor arg for the name of the folder. Charm authors can choose whether to expose it as a config option

We need to remember that if this config option is changed during the lifetime of a charm in Grafana we need to move the dashboards to the new folder.

Reply:
Added.